rating/peersRatingHandler.go

```diff
@@ -129,10 +183,88 @@ func (prh *peersRatingHandler) getOldRating(pid core.PeerID) (int32, bool) {
 	return defaultRating, found
 }
 
-func (prh *peersRatingHandler) updateRatingIfNeeded(pid core.PeerID, updateFactor int32) {
-	defer prh.updateMetrics()
+func (prh *peersRatingHandler) markInactivePIDsForRemoval(receivedPIDs map[string]struct{}) {
+	removalTimestamp := prh.getTimeHandler().Add(prh.timeWaitingForReconnection).Unix()
+	topRatedPIDs := prh.topRatedCache.Keys()
+	for _, pid := range topRatedPIDs {
+		_, isPIDStillActive := receivedPIDs[string(pid)]
+		alreadyMarked := prh.markedForRemovalCache.Has(pid)
+		if !isPIDStillActive && !alreadyMarked {
+			prh.markedForRemovalCache.Put(pid, removalTimestamp, int64Size)
+		}
+	}
+	badRatedPIDs := prh.badRatedCache.Keys()
+	for _, pid := range badRatedPIDs {
+		_, isPIDStillActive := receivedPIDs[string(pid)]
+		alreadyMarked := prh.markedForRemovalCache.Has(pid)
+		if !isPIDStillActive && !alreadyMarked {
+			prh.markedForRemovalCache.Put(pid, removalTimestamp, int64Size)
+		}
+	}
+}
+
+func (prh *peersRatingHandler) sweepCachersLoop(ctx context.Context) {
+	timer := time.NewTimer(prh.timeBetweenCachersSweep)
+	defer timer.Stop()
 
-	oldRating, found := prh.getOldRating(pid)
+	for {
+		timer.Reset(prh.timeBetweenCachersSweep)
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+		}
+		prh.sweepCachers()
+	}
+}
+
+func (prh *peersRatingHandler) sweepCachers() {
+	prh.mut.Lock()
```

Comment:
I think the mutex operation can be moved inside removePIDFromCachers function

Reply:
switching markedForRemovalCache to a map would require this lock to stay on this line in order to protect that map as well